### PR TITLE
Make it possible to implement your own network layer #58

### LINKED ISSUE #58: Make it possible to implement your own network layer
Just like the iOS project (https://github.com/parley-messaging/ios-library/issues/77), we would like to be able to do all the network traffic with our own implementation. 

To accomplice this, the dependencies `Retrofit`, `Glide` and `Trustkit` should be moved to new module, let's call it `Parley-Network`. This module would then be optional to include as a dependency. The rest of the SDK should be loosely coupled with the networking layer, so it should only interact with it via interfaces.

This setup would allow us to do all api requests with our existing network layer instead of using custom interceptors to allow interaction with our secured backend. This would bring us the following benefits:
- It would be less maintenance in the long term.
- This would also allow us to fix an issue with loading images, since they don't add the correct http headers at the moment.
- It would allow us to correctly handle networking/security errors, since we can convert them to the format that the SDK expects. This is currently not possible with interceptors.

### PR DESCRIPTION
**Pull request by mennovogel**
This fixes #58

I've made it optional to use Retrofit and its configurations (OkHttp and Gson). 

Just like it was already possible to implement your own version of the ParleyLaunchCallback with a default implementation of DefaultParleyLaunchCallback, I've used the same logic to implement your own version of the newly added NetworkConfig interface. A DefaultNetworkConfig is included and is used by default, so no changes are needed on a version update of Parley.

Let me know if you have questions or other comments.

### CODE REVIEW

File: parley/src/main/java/nu/parley/android/data/net/service/DeviceService.kt
Comment (mennovogel) [inline parley/src/main/java/nu/parley/android/data/net/service/DeviceService.kt:1]: I've converted the Retrofit Services interfaces to Kotlin.

File: parley/src/main/java/nu/parley/android/ParleyNetwork.java [hunk 1]
```diff
@@ -20,12 +17,11 @@ public final class ParleyNetwork {
 
     public final String url;
     public final Map<String, String> headers;
-    @Nullable
-    public Interceptor interceptor;
```
[the remark this answers is not in the document]
Comment (alexkok) [reply parley/src/main/java/nu/parley/android/ParleyNetwork.java:24]: Agreed, good idea 👍

**Review by alexkok — CHANGES_REQUESTED** [review]

**Comment by alexkok** [comment]
With this implementation, the developer would need to implement all repositories by their interfaces. 
Feels like a bit too much of a requirement to do so. It would be more neat if we just interface the Network Client instead. 

Doing so would maybe result in a change in how we use Retrofit, but that's alright. Interfacing only the network client would allow intercepting requests by the developer. And doesn't require them to implement all the repositories we have.

For example, in iOS we have:
```
class CustomNetworkSession : ParleyNetworkSession {
    func request(_ url: URL, data: Data?, method: ParleyHTTPRequestMethod, headers: [String : String], completion: @escaping (Result<ParleyHTTPDataResponse, ParleyHTTPErrorResponse>) -> Void) {
        // ...
    }
    
    func upload(data: Data, to url: URL, method: ParleyHTTPRequestMethod, headers: [String : String], completion: @escaping (Result<ParleyHTTPDataResponse, ParleyHTTPErrorResponse>) -> Void) {
        // ...
    }
}
```

After interfacing network client, we would like to align terminology with iOS (NetworkSession etc.) -- Or if we have ideas for better naming, we might as well adjust iOS to match these again in a backwards compatible way. We'll look at naming later.

What do you think about this? Would interfacing the network client work as well instead of interfacing the repositories?

_Sidenote: Also, Parley keeps the dependency on retrofit like this. But the good thing is that it's not a breaking change because of that, so that's great, that's good for now. We'll likely separate that when a v4 comes (note that v4 is not planned yet)._

File: parley/src/main/java/nu/parley/android/util/DeviceUtil.java
[the remark this answers is not in the document]
Comment (alexkok) [reply parley/src/main/java/nu/parley/android/util/DeviceUtil.java:1]: Perfect, yeah that is nice anyway, to separate that logic

**Comment by mennovogel** [comment]
> With this implementation, the developer would need to implement all repositories by their interfaces.
> Feels like a bit too much of a requirement to do so. It would be more neat if we just interface the Network Client instead.
> 
> Doing so would maybe result in a change in how we use Retrofit, but that's alright. Interfacing only the network client would allow intercepting requests by the developer. And doesn't require them to implement all the repositories we have.
> 

I've looked at the iOS implementation, but I didn't think it was possible without a lot of refactoring and/or removing the usage of Retrofit. Now that I think more about it, we might be able to alter the usage of it and use Retrofit only for passing Strings, without the json mapping. I'll look into that.

> After interfacing network client, we would like to align terminology with iOS (NetworkSession etc.) 

I'll check the naming and adjust it.

**Comment by mennovogel** [comment]
As I proposed, I've changed the way we use Retrofit, now it only passes Strings and the json parsing is always done by the Parley library. This makes it easier to implement the ParleyNetworkSession interface.

**Comment by mennovogel** [comment]
@alexkok
Could you have another look please?

**Comment by alexkok** [comment]
Thanks @mennovogel ! Looks very good 👍 

FYI: Found another issue, but that is not related to these changes. So fixing that as well now before creating a new version.

**Review by alexkok — APPROVED** [review]

**Comment by alexkok** [comment]
Released in [3.11.0](https://github.com/parley-messaging/android-library/releases/tag/3.11.0) 🚀

**Pull request merged**